warehouse/jobs/runner.go

```diff
@@ -0,0 +1,373 @@
+	for {
+		asyncWhJob.log.Info("WH-Jobs: Scanning for waiting async job")
+		select {
+		case <-ctx.Done():
+			asyncWhJob.log.Info("WH-Jobs: Stopping AsyncJobRunner")
+			return
+		case <-time.After(10 * time.Second):
+
+		}
+
+		asyncjobpayloads, err := asyncWhJob.getPendingAsyncJobs(ctx)
+		if err != nil {
+			return
+		}
+
+		if len(asyncjobpayloads) > 0 {
+			asyncWhJob.log.Info("WH-Jobs: Got pending wh async jobs")
+			asyncWhJob.log.Infof("WH-Jobs: Number of async wh jobs left = %d\n", len(asyncjobpayloads))
+			notifierClaims, err := getMessagePayloadsFromAsyncJobPayloads(asyncjobpayloads)
+
+			if err != nil {
+				asyncWhJob.log.Errorf("Error converting the asyncJobType to notifier payload %s ", err)
+				asyncWhJob.updateMultipleAsyncJobs(&asyncjobpayloads, WhJobFailed, err.Error())
+				return
+			}
+			messagePayload := pgnotifier.MessagePayload{
+				Jobs:    notifierClaims,
+				JobType: "async_job",
+			}
+			schema := warehouseutils.SchemaT{}
+			ch, err := asyncWhJob.pgnotifier.Publish(messagePayload, &schema, 100)
+			if err != nil {
+				asyncWhJob.updateMultipleAsyncJobs(&asyncjobpayloads, WhJobFailed, err.Error())
+				return
+			}
+			asyncWhJob.updateMultipleAsyncJobs(&asyncjobpayloads, WhJobExecuting, "")
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go func() {
+				//Is this correct or should I be passing ch as a variable to the anonymous go routine instead?
+				responses := <-ch
+				asyncWhJob.log.Info("Response recieved from the pgnotifier track batch")
+				var successfulAsyncJobs []AsyncJobPayloadT
+				for _, resp := range responses {
+					var jobID pgnotifier.AsyncOutput
+					err = json.Unmarshal(resp.Output, &jobID)
+					if err != nil {
+						asyncWhJob.log.Errorf("Unable unmarshal output %s", err)
+						//Will skip to next response, leaving the table row in executing stage.
+						// We need to program retry strategy, timeout strategy for each row in the asyncJobtable
+						continue
+					}
+					ind := getSinglePayloadFromBatchPayloadByTableName(&asyncjobpayloads, jobID.TableName)
+					if ind == -1 {
+						//Continues with the next payload leaving this row in the executing stage
+						asyncWhJob.log.Errorf("tableName is not found while updating the response after completing the trackasyncJob %s", jobID.TableName)
+						continue
+					}
+					if resp.Status == pgnotifier.AbortedState {
+						pkgLogger.Errorf("[WH]: Error in running async task %v", resp.Error)
+						sampleError := fmt.Errorf(resp.Error)
+						asyncWhJob.updateSingleAsyncJob(&asyncjobpayloads[ind], WhJobAborted, sampleError.Error())
+						continue
+					}
+					if resp.Status == pgnotifier.FailedState {
+						pkgLogger.Errorf("[WH]: Error in running async task %v", resp.Error)
+						sampleError := fmt.Errorf(resp.Error)
+						asyncWhJob.updateSingleAsyncJob(&asyncjobpayloads[ind], WhJobFailed, sampleError.Error())
+						continue
+					}
+					successfulAsyncJobs = append(successfulAsyncJobs, asyncjobpayloads[ind])
```

Comment:
Instead of handling separately lets wrap them with some struct and make one single db call at the end.

Reply:
Sure